src/library/compiler/ir_interpreter.cpp

```diff
@@ -0,0 +1,745 @@
+/*
+Copyright (c) 2019 Microsoft Corporation. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+
+Author: Sebastian Ullrich
+
+A simple interpreter for evaluating λRC IR code.
+
+Motivation
+==========
+
+Even with a JIT compiler, we still have a need for a simpler interpreter on platforms LLVM JIT does not support (i.e.
+WebAssembly). Because this is mostly an edge case, we strive for simplicity instead of performance and thus reuse the
+existing compiler IR instead of inventing something like a new bytecode format.
+
+Implementation
+==============
+
+The interpreter mainly consists of a homogeneous stack of boxed `object *` values. IR variables are mapped to stack
+slots by adding the current base pointer to the variable index. Further stacks are used for storing join points and call
+stack metadata. The interpreted IR is taken directly from the environment. Whenever possible, we try to switch to native
+code by checking for the mangled symbol via dlsym/GetProcAddress, which is also how we can call external functions (but
+only if the file declaring them has already been compiled). We always call the "boxed" versions of native functions,
+which have a (relatively) homogeneous ABI that we can use without runtime code generation; see also `call/lookup_symbol`
+below.
+
+*/
+#include <string>
+#include <vector>
+#include <dlfcn.h>
+#include "runtime/flet.h"
+#include "runtime/apply.h"
+#include "library/trace.h"
+#include "library/compiler/ir.h"
+#include "util/option_ref.h"
+#include "util/array_ref.h"
+#include "util/nat.h"
+
+namespace lean {
+namespace ir {
+// C++ wrappers of Lean data types
+
+typedef object_ref lit_val;
+typedef object_ref ctor_info;
+
+// TODO(Sebastian): move
+template<class T>
+inline T const & cnstr_get_ref_t(object_ref const & o, unsigned i) {
+    static_assert(sizeof(T) == sizeof(object_ref), "unexpected object wrapper size");
+    return *reinterpret_cast<T const *>(&cnstr_get_ref(o.raw(), i));
+}
```

Comment:
@leodemoura Does this look sensible? I'll move it to `object_ref.h` then.

Reply:
I found this line weird. Please use  
```cpp
    return static_cast<T const &>(cnstr_get_ref(o.raw(), i));
```
instead.

Reply:
changed & moved